### PR TITLE
usbus: Add automated test

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds a basic test for validating the USB enumeration sequence as performed by a host. Responses from the USBUS stack are validated by using a mock usbdev device which is faking the interaction.

There are `DEBUG` statement in the test which can be enabled to track during which request the stack caused a failed assertion.

A number of issues with USBUS were found during the writing of this test (and during the writing of the CDC ACM functionality) which are fixed in the first commit.

### Testing procedure

Run test, check output.

### Issues/PRs references

Required for #11085

### CODE REVIEW

File: tests/usbus/Makefile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+USB_VID ?= 1209
+USB_PID ?= 0001
+
+CFLAGS += -DUSB_CONFIG_VID=0x$(USB_VID) -DUSB_CONFIG_PID=0x$(USB_PID)
+
+include $(RIOTBASE)/Makefile.include
+
+.PHONY: usb_id_check
+usb_id_check:
+	@if [ $(USB_VID) = $(DEFAULT_VID) -o $(USB_PID) = $(DEFAULT_PID) ] ; then \
```
Comment (miri64) [inline tests/usbus/Makefile]: `-o` is a bash-ism (and Murdock complains about it.
```suggestion
	@if [ $(USB_VID) = $(DEFAULT_VID) ] || [ $(USB_PID) = $(DEFAULT_PID) ] ; then \
```

File: tests/usbus/main.c [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    TEST_REQ_PHASE_START_GET,
+    TEST_REQ_PHASE_START_SET,
+    TEST_REQ_PHASE_OUTDATA,
+    TEST_REQ_PHASE_INDATA,
+    TEST_REQ_PHASE_OUTACK,
+    TEST_REQ_PHASE_INACK,
+} test_req_phase_t;
+
+testphase_t phase = TESTPHASE_START;
+test_req_phase_t req_phase = TESTPHASE_START;
```
Comment (miri64) [inline tests/usbus/main.c]:
```suggestion
test_req_phase_t req_phase = TEST_REQ_PHASE_IDLE;
```
?

File: tests/usbus/Makefile (hunk 1 above)
Comment (miri64) [reply tests/usbus/Makefile]: See also [POSIX recommendations](https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/utilities/test.html#tag_20_128_16)

File: tests/usbus/main.c (hunk 2 above)
Comment (bergzand) [reply tests/usbus/main.c]: Oh wow… :flushed:

File: tests/usbus/Makefile (hunk 1 above)
Comment (bergzand) [reply tests/usbus/Makefile]: [Sorry](https://github.com/RIOT-OS/RIOT/pull/12146)

Comment (aabadie) [reply tests/usbus/Makefile]: Maybe just replace [this line](https://github.com/RIOT-OS/RIOT/blob/master/.murdock#L1) with `#!/usr/bin/env bash` ?
I'm wondering why `/bin/sh` is still used instead of bash and what's wrong with bashing.

Comment (miri64) [reply tests/usbus/Makefile]: > […] and what's wrong with ~bashing~ (oups ;) ) bash-ism.

~~One dependency less~~ (ok since we use already plenty of them this argument is invalid) and assuming that bash is the default shell (which it is often, but not always ;-)).

File: tests/usbus/main.c (hunk 2 above)
Comment (bergzand) [reply tests/usbus/main.c]: Fixed

Comment (bergzand) [reply tests/usbus/main.c]: Never mind, took a second attempt to actually fix it. Time to go to bed.

File: tests/usbus/Makefile (hunk 1 above)
Comment (bergzand) [reply tests/usbus/Makefile]: I don't mind changing this as I prefer to keep things POSIX compatible where possible

Comment (bergzand) [reply tests/usbus/Makefile]: Changed to the POSIX alternative

File: sys/usb/usbus/usbus_control.c [hunk 3]
```diff
@@ -213,21 +212,19 @@ static int _recv_dev_setup(usbus_t *usbus, usb_setup_t *pkt)
             case USB_SETUP_REQ_SET_ADDRESS:
                 DEBUG("usbus_control: Setting address\n");
                 usbus->addr = (uint8_t)pkt->value;
-                res = 0;
+                res = 1;
```
Comment (dylad) [inline sys/usb/usbus/usbus_control.c]: Any reason to change this ? Usually we consider 0 to be OK and < 0 not OK

File: sys/usb/usbus/usbus_control.c [hunk 4]
```diff
@@ -307,6 +307,17 @@ static void _usbus_config_ep0(usbus_control_handler_t *ep0_handler)
     usbdev_ep_ready(ep0_handler->out, 0);
 }
 
+uint8_t *usbus_control_get_out_data(usbus_t *usbus, size_t *len)
+{
+    assert((int)usbus->state == (int)USBUS_CONTROL_REQUEST_STATE_OUTDATA);
```
Comment (dylad) [inline sys/usb/usbus/usbus_control.c]: Is it desireable to mix up `usbus_state_t` with `usbus_setuprq_state_t` ?

File: tests/usbus/main.c [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    if (req_phase == TEST_REQ_PHASE_IDLE) {
+        DEBUG("[ep esr]: Done with the request\n");
+        /* signal USBDEV_EVENT_ESR to call _test_sequence */
+        dev->usbdev.cb(&dev->usbdev, USBDEV_EVENT_ESR);
+    }
+    ep->ep.dev->epcb(&ep->ep, USBDEV_EVENT_TR_COMPLETE);
+}
+
+/* Handles Control endpoint validation from the usbdev side  */
+/* Called as part of the endpoint ready function             */
```
Comment (dylad) [inline tests/usbus/main.c]: put these comments together

File: tests/usbus/usbdev_mock.c [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+{
+    usbdev_mock_ep_t *testep = (usbdev_mock_ep_t *)ep;
+
+    (void)max_len;
+    switch (opt) {
+        case USBOPT_EP_AVAILABLE:
+            *((size_t *)value) = testep->available;
+            return sizeof(size_t);
+        default:
+            break;
```
Comment (dylad) [inline tests/usbus/usbdev_mock.c]: Please add a DEBUG statement

File: tests/usbus/usbdev_mock.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                                       usbdev_mock_ep_t *ep,
+                                       size_t len);
+
+/**
+ * @brief usbdev mock device
+ */
+struct usbdev_mock {
+    usbdev_t usbdev;                    /**< Generic usbdev device          */
+    usbdev_mock_ep_t in[1];             /**< IN endpoints                   */
+    usbdev_mock_ep_t out[1];            /**< OUT endpoints                  */
```
Comment (dylad) [inline tests/usbus/usbdev_mock.h]: why an array with one element ??

Comment (bergzand) [reply tests/usbus/usbdev_mock.h]: To easily extend the test as soon as more endpoints are required for the test. Prevents us from having to change all the accesses to include array indices

File: sys/usb/usbus/usbus_control.c (hunk 4 above)
Comment (bergzand) [reply sys/usb/usbus/usbus_control.c]: No, definitely not

File: sys/usb/usbus/usbus_control.c (hunk 3 above)
Comment (bergzand) [reply sys/usb/usbus/usbus_control.c]: This must follow the same return codes from the `control_handler` function:
 - Negative on error, results in USB stall
 - Zero when it is not handled, results in an USB nack if nothing handles the request
 - Positive for the length, which is useful for USB IN transactions, but doesn't really mean anything for OUT transactions where the host decides the amount of data.

This last case could be documented a bit better for OUT transactions

Comment (bergzand) [reply sys/usb/usbus/usbus_control.c]: It is used to determine the response to the transfer here:
https://github.com/RIOT-OS/RIOT/blob/c4b103306c81ce84eb3bceab80508839a6900172/sys/usb/usbus/usbus_control.c#L277-L294

**Comment by bergzand** [comment]
I've added the python code required for `make test`, not sure where that went on my initial push

File: sys/usb/usbus/usbus_control.c (hunk 4 above)
Comment (bergzand) [reply sys/usb/usbus/usbus_control.c]: Fixed

File: tests/usbus/main.c (hunk 5 above)
Comment (bergzand) [reply tests/usbus/main.c]: Fixed

File: tests/usbus/usbdev_mock.c (hunk 6 above)
Comment (bergzand) [reply tests/usbus/usbdev_mock.c]: Added

File: tests/usbus/usbdev_mock.h (hunk 7 above)
Comment (dylad) [reply tests/usbus/usbdev_mock.h]: Then you should consider to use a `#define` ;)

**Comment by dylad** [comment]
Just give it a go as I was playing with SAME54-XPRO

Current terminal output if ENABLE_DEBUG (0)
```
2019-09-18 13:33:01,251 - INFO # main(): This is RIOT! (Version: 2019.04-devel-2609-g52a35-pr/usbus/auto_tests)
2019-09-18 13:33:01,252 - INFO # .
2019-09-18 13:33:01,253 - INFO # OK (1 tests)
```
This is not very meaningful, right ? :)

**Comment by bergzand** [comment]
> This is not very meaningful, right ? :)

I didn't think it was necessary to have more verbose output by default. Similar to how most automated tests using embunit don't really have any descriptive output.

File: sys/include/usb/usbus/control.h [hunk 8]
```diff
@@ -123,6 +128,18 @@ void usbus_control_slicer_ready(usbus_t *usbus);
...
+ * @brief Retrieve the data from the OUT pipe of the control endpoint
+ *
+ * @pre usbus->state == USBUS_SETUPRQ_OUTDATA
+ *
+ * @param[in]   usbus       USBUS context
+ * @param[out]  len         Length of the data part
+ *
+ * @return                  pointer to the data buffer
+ */
+uint8_t *usbus_control_get_out_data(usbus_t *usbus, size_t *len);
```
Comment (bergzand) [inline sys/include/usb/usbus/control.h]: Just to be honest, this function is not strictly necessary for this PR, it is not tested in the automated test, and can be split off to a separate PR if desired.

File: tests/usbus/usbdev_mock.h (hunk 7 above)
Comment (bergzand) [reply tests/usbus/usbdev_mock.h]: Done!

File: tests/usbus/main.c [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+        case TEST_REQ_PHASE_START_SET:
+            return "start set";
+        case TEST_REQ_PHASE_OUTDATA:
+            return "OUT data";
+        case TEST_REQ_PHASE_INDATA:
+            return "IN data";
+        case TEST_REQ_PHASE_OUTACK:
+            return "OUT Ack";
+        case TEST_REQ_PHASE_INACK:
+            return "IN Ack";
```
Comment (dylad) [inline tests/usbus/main.c]: missing default statement

File: tests/usbus/main.c [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+        case TEST_REQ_PHASE_START_SET:
+            _validate_set_req_start(dev, ep, len);
+            break;
+        case TEST_REQ_PHASE_OUTACK:
+            _validate_out_ack(dev, ep);
+            break;
+        case TEST_REQ_PHASE_INACK:
+            _validate_in_ack(dev, ep, len);
+            break;
+        default:
```
Comment (dylad) [inline tests/usbus/main.c]: please add a DEBUG statement

File: tests/usbus/tests/01-run.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2016 Kaspar Schleiser <kaspar@schleiser.de>
+# Copyright (C) 2016 Takuo Yonezawa <Yonezawa-T2@mail.dnp.co.jp>
```
Comment (dylad) [inline tests/usbus/tests/01-run.py]: Hmmmmm, really ?

File: tests/usbus/usbdev_mock.c [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+static const usbdev_driver_t testdriver;
+
+static usbdev_mock_t *_ep2dev(usbdev_ep_t *ep)
+{
+    return (usbdev_mock_t *)ep->dev;
+}
+
+void usbdev_init_lowlevel(void)
+{}
```
Comment (dylad) [inline tests/usbus/usbdev_mock.c]:
```suggestion
{

}
```

File: tests/usbus/usbdev_mock.c [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+    usbdev_mock.ep_esr_cb = ep_esr_cb;
+    usbdev_mock.ready_cb = ready_cb;
+}
+
+static void _init(usbdev_t *usbdev)
+{
+    usbdev_mock_t *dev = (usbdev_mock_t *)usbdev;
+
+    /* Init fake usbdev */
+
```
Comment (dylad) [inline tests/usbus/usbdev_mock.c]: Missing call or forgot to remove comment ?

File: tests/usbus/usbdev_mock.h [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * @brief usbdev mock device forward declaration
+ */
+typedef struct usbdev_mock usbdev_mock_t;
+
+/**
+ * @brief usbdev mock device callback for esr event checking
+ *
+ * @param dev   usbdev mock device state
```
Comment (dylad) [inline tests/usbus/usbdev_mock.h]: You should change this description
Same for all functions within this file using dev arg

File: tests/usbus/tests/01-run.py (hunk 11 above)
Comment (bergzand) [reply tests/usbus/tests/01-run.py]: Yes, really, I copied this file from one of the other tests. I didn't change the file significantly so I don't think I should claim copyright.

File: tests/usbus/main.c (hunk 9 above)
Comment (bergzand) [reply tests/usbus/main.c]: Intentionally missing, in that case the empty string at [line 87](https://github.com/RIOT-OS/RIOT/pull/12267/files/a33c46f3f7683ad6b8e93bdf954b942da5dffed8#diff-e989200794ef37dacd03b000cad21afcR87) will be returned

Comment (bergzand) [reply tests/usbus/main.c]: I can change the empty string to something like `"unknown control request state"` though.

File: tests/usbus/main.c (hunk 10 above)
Comment (bergzand) [reply tests/usbus/main.c]: The `default` case here is actively used with:
 - `TEST_REQ_PHASE_IDLE`
 - `TEST_REQ_PHASE_OUTDATA`
 - `TEST_REQ_PHASE_INDATA`

I don't assert anything at the moment for those stages.

I can add a comment there to clarify this or I can add a debug statement if you have a suggestion that is generic enough supply sensible information for these 3 cases.

Comment (bergzand) [reply tests/usbus/main.c]: I can also split out these cases and add a simple debug statement for each, which is what I prefer as that would make the test easier to follow when `DEBUG` is enabled (I think).

Any preference?

File: tests/usbus/main.c (hunk 9 above)
Comment (bergzand) [reply tests/usbus/main.c]: Fixed

File: tests/usbus/usbdev_mock.c (hunk 12 above)
Comment (bergzand) [reply tests/usbus/usbdev_mock.c]: Fixed

File: tests/usbus/usbdev_mock.c (hunk 13 above)
Comment (bergzand) [reply tests/usbus/usbdev_mock.c]: Removed

File: tests/usbus/usbdev_mock.h (hunk 14 above)
Comment (bergzand) [reply tests/usbus/usbdev_mock.h]: Modified

File: tests/usbus/main.c (hunk 10 above)
Comment (dylad) [reply tests/usbus/main.c]: a comment should be enough, there is no need to spam output if it's heavily used.

Comment (bergzand) [reply tests/usbus/main.c]: And added a comment :)

**Comment by bergzand** [comment]
> ACK
> changes are good, tested the provided test + cdc_ecm and it works
> please squash

Squashed, thanks for testing!

**Comment by dylad** [comment]
Here we go !

**Comment by bergzand** [comment]
Thanks!
